interpreter/syntax/types.ml

```diff
@@ -102,222 +127,166 @@ let string_of_name n =
 
-let string_of_module_type : module_type -> string = function
-  | ModuleT (dts, its, ets) ->
+let string_of_module_type = function
+  | ModuleT (its, ets) ->
     String.concat "" (
-      List.mapi (fun i dt -> "type " ^ string_of_int i ^ " = " ^ string_of_def_type dt ^ "\n") dts @
       List.map (fun it -> "import " ^ string_of_import_type it ^ "\n") its @
       List.map (fun et -> "export " ^ string_of_export_type et ^ "\n") ets
     )
 
-(* Dynamic Types *)
-
-type type_addr += Addr of def_type Lib.Promise.t
-
-let unwrap = function
-  | Addr p -> p
-  | _ -> assert false
-
-let alloc_uninit () = Addr (Lib.Promise.make ())
-let init x dt = Lib.Promise.fulfill (unwrap x) dt
-let alloc dt = let x = alloc_uninit () in init x dt; x
-let def_of x = Lib.Promise.value (unwrap x)
-
-let () = string_of_addr' :=
-  let inner = ref false in
-  fun x ->
-    if !inner then "..." else
-    ( inner := true;
-      try
-        let s = string_of_def_type (def_of x) in
-        inner := false; "(" ^ s ^ ")"
-      with exn -> inner := false; raise exn
-    )
-
+(* Substitution *)
 
-(* Instantiation *)
+type subst = var -> heap_type
 
-let dyn_var_type c = function
-  | Stat x -> Dyn (Lib.List32.nth c x)
-  | Dyn a -> assert false
+let subst_num_type s t = t
 
-let dyn_num_type c = function
-  | t -> t
+let subst_vec_type s t = t
 
-let dyn_vec_type c = function
-  | t -> t
-
-let dyn_heap_type c = function
+let subst_heap_type s = function
   | FuncHT -> FuncHT
   | ExternHT -> ExternHT
-  | DefHT x -> DefHT (dyn_var_type c x)
+  | VarHT x -> s x
+  | DefHT ht -> DefHT ht  (* assume closed *)
   | BotHT -> BotHT
 
-let dyn_ref_type c = function
-  | (nul, t) -> (nul, dyn_heap_type c t)
+let subst_ref_type s = function
+  | (nul, t) -> (nul, subst_heap_type s t)
 
-let dyn_val_type c = function
-  | NumT t -> NumT (dyn_num_type c t)
-  | VecT t -> VecT (dyn_vec_type c t)
-  | RefT t -> RefT (dyn_ref_type c t)
+let subst_val_type s = function
+  | NumT t -> NumT (subst_num_type s t)
+  | VecT t -> VecT (subst_vec_type s t)
+  | RefT t -> RefT (subst_ref_type s t)
   | BotT -> BotT
 
-let dyn_result_type c = function
-  | ts -> List.map (dyn_val_type c) ts
-
-let dyn_func_type c = function
-  | FuncT (ts1, ts2) -> FuncT (dyn_result_type c ts1, dyn_result_type c ts2)
+let subst_result_type s = function
+  | ts -> List.map (subst_val_type s) ts
 
-let dyn_cont_type c = function
-  | ContT x -> ContT (dyn_var_type c x)
+let subst_func_type s = function
+  | FuncT (ts1, ts2) -> FuncT (subst_result_type s ts1, subst_result_type s ts2)
 
-let dyn_def_type c = function
-  | DefFuncT ft -> DefFuncT (dyn_func_type c ft)
-  | DefContT x  -> DefContT (dyn_cont_type c x)
+let subst_cont_type s = function
+  | ContT ft -> ContT (subst_heap_type s ft)
 
-let dyn_local_type c = function
-  | LocalT (init, t) -> LocalT (init, dyn_val_type c t)
+let subst_def_type s = function
+  | DefFuncT ft -> DefFuncT (subst_func_type s ft)
+  | DefContT ct -> DefContT ct
```

Comment:
I think the bug is here.
```suggestion
  | DefContT ct -> DefContT (subst_cont_type s ct)
```

Reply:
Ah thanks! I had starred blind at it for a while!